_posts/2019-05-10-why-projects-need-codenames.markdown

```diff
@@ -0,0 +1,48 @@
+---
+layout: epic
+title: Why Projects Need Code Names
+date: 2019-05-10
+author: [joey]
+categories: [engineering, git, github]
+---
+
+Before I joined Artsy, I worked at companies where software projects tended to have meaningful, predictable names. If we were building a system for flagging media uploads, it might be called `media-review`. In many cases, our code repositories' names matched the main product's branding or even the company's name. Life was simple and there was no risk of ambiguity.
+
+At Artsy, our systems have peculiar code names like _Gravity_, _Pulse_, and _Vortex_. There's a persistent learning curve as you contribute to different repositories or as new services get created. Numerous times, I've wondered: are code names worth the trouble?
+
+<!-- more -->
+
+![](/images/2019-05-10-why-projects-need-codenames/github_projects.png)
+
+To be clear, _any_ project naming scheme works in small quantities. Personal projects or libraries for public release should probably just be named for clarity. Code names start to offer benefits as a team grows and a product evolves over the longer term:
+
+Code names embrace that **we don't necessarily know the fit or scope of a system when it's first conceived**. To _not_ employ code names would require accurately predicting a system's eventual function. A project named `artsy-admin` starts to grate when we make the architectural decision to split apart management utilities or introduce more fine-grained administrative roles. An `artwork-taxonomy` label loses meaning when that system expands to include less structured tags
+
+**Branding shifts, products pivot, and companies merge.** Project code names introduce a level of indirection between what engineers build and the labels, URLs, or brands that end users experience. Over a long enough period, these external or superficial changes _will_ happen and risk introducing confusion or just subtle misalignment between form and function. As with software design in general, abstractions can be a powerful tool to [separate concerns](https://en.wikipedia.org/wiki/Separation_of_concerns). Ask yourself: what code changes might be required if your company were to spin off your product or be merged into an acquirer?
+
+**Language is important**, and not just for communication but for shaping our thoughts and assumptions. We frequently find ourselves debating which system should serve as the authority for a given domain model or where to implement a new feature, and project names that overlap with these topics (e.g., `search`, `images`, `suggestions`) would predispose us to certain decisions. Code names free us to focus on the architectural and organizational merits instead.
+
+We like to **have fun at work** and it's more joyful to proclaim that "Torque is in the wild" than "data-sync has been deployed." Over time these names gain mythologies and personalities within the team and organization.
```

Comment:
Was going to comment about the period inside the quotation marks, but then I learned that this is another Simplified English thing 🤷‍♂